src/mam/mod_mam_muc_rdbms_arch.erl

```diff
@@ -137,47 +136,35 @@ archive_size(Size, Host, RoomID, _RoomJID) when is_integer(Size) ->
     mongoose_rdbms:result_to_integer(BSize).
 
 -spec archive_message(_Result, jid:server(), MessID :: mod_mam:message_id(),
-                      RoomID :: mod_mam:archive_id(), _LocJID :: jid:jid(),
+                      RoomID :: mod_mam:archive_id(), BareSenderJID :: jid:jid(),
                       _RemJID :: jid:jid(),
-                      _SrcJID :: jid:jid(), incoming, Packet :: packet()) -> ok.
-archive_message(_Result, Host, MessID, RoomID,
-                _LocJID=#jid{},
-                _RemJID=#jid{},
-                _SrcJID=#jid{lresource=FromNick}, incoming, Packet) ->
+                      UserRoomJID :: jid:jid(), incoming, Packet :: packet()) -> ok.
+archive_message(_Result, Host, MessID, RoomID, BareSenderJID = #jid{},
+                _RemJID = #jid{}, UserRoomJID = #jid{}, incoming, Packet) ->
     try
-        archive_message_unsafe(Host, MessID, RoomID, FromNick, Packet)
+        Row = prepare_message(Host, MessID, RoomID, BareSenderJID, UserRoomJID, Packet),
+        {updated, 1} = mod_mam_utils:success_sql_execute(Host, insert_mam_muc_message, Row),
+        ok
     catch _Type:Reason ->
             ?ERROR_MSG("event=archive_message_failed mess_id=~p room_id=~p "
                        "from_nick=~p reason='~p' stacktrace=~p",
-                       [MessID, RoomID, FromNick, Reason, erlang:get_stacktrace()]),
+                       [MessID, RoomID, UserRoomJID#jid.lresource, Reason, erlang:get_stacktrace()]),
             {error, Reason}
     end.
 
--spec archive_message_unsafe(jid:server(), mod_mam:message_id(), mod_mam:archive_id(),
-                             FromNick :: jid:user(), packet()) -> ok.
-archive_message_unsafe(Host, MessID, RoomID, FromNick, Packet) ->
-    Row = prepare_message(Host, MessID, RoomID, FromNick, Packet),
-    {updated, 1} = mod_mam_utils:success_sql_execute(Host, insert_mam_muc_message, Row),
-    ok.
-
--spec prepare_message(jid:server(), mod_mam:message_id(), mod_mam:archive_id(),
-                      _LocJID :: jid:jid(), _RemJID :: jid:jid(),
-                      _SrcJID :: jid:jid(), incoming, packet()) -> list().
-prepare_message(Host, MessID, RoomID,
-                _LocJID=#jid{},
-                _RemJID=#jid{},
-                _SrcJID=#jid{lresource=FromNick}, incoming, Packet) ->
-    prepare_message(Host, MessID, RoomID, FromNick, Packet).
-
-prepare_message(Host, MessID, RoomID, FromNick, Packet) ->
+-spec prepare_message(Host :: jid:server(), MessID :: mod_mam:message_id(),
+                      RoomID :: mod_mam:archive_id(), BareSenderJID :: jid:jid(),
+                      UserRoomJID :: jid:jid(), Packet :: packet()) -> list().
```

Comment:
the returned list is not just any list, would it make sense to try to be more specific about returned type?

Reply:
Sure, why not.